adapters/AppLovin/AppLovinAdapter/GADMAdapterAppLovin.m

```diff
@@ -15,296 +15,298 @@
-  self.interstitialDelegate = nil;
-  self.bannerDelegate = nil;
+  _connector = nil;
+  _interstitialDelegate = nil;
+  _bannerDelegate = nil;
 
-  self.interstitial.adDisplayDelegate = nil;
-  self.interstitial.adVideoPlaybackDelegate = nil;
+  _interstitial.adDisplayDelegate = nil;
+  _interstitial.adVideoPlaybackDelegate = nil;
 
-  self.adView.adLoadDelegate = nil;
-  self.adView.adDisplayDelegate = nil;
-  self.adView.adEventDelegate = nil;
+  _adView.adLoadDelegate = nil;
+  _adView.adDisplayDelegate = nil;
+  _adView.adEventDelegate = nil;
 }
 
-#pragma mark - GAD Ad Network Protocol Interstitial Methods
+#pragma mark - GADMAdNetworkAdapter Protocol Interstitial Methods
 
 - (void)getInterstitial {
-  id<GADMAdNetworkConnector> strongConnector = self.connector;
-  self.zoneIdentifier =
-      [GADMAdapterAppLovinUtils retrieveZoneIdentifierFromConnector:strongConnector];
+  id<GADMAdNetworkConnector> strongConnector = _connector;
+
+  if (!_sdk) {
+    NSError *error = GADMAdapterAppLovinErrorWithCodeAndDescription(
+        kGADErrorMediationAdapterError, @"Failed to retrieve SDK instance.");
+    [strongConnector adapter:self didFailAd:error];
+    return;
+  }
+
+  _zoneIdentifier = [GADMAdapterAppLovinUtils zoneIdentifierForConnector:strongConnector];
 
   // Unable to resolve a valid zone - error out
-  if (!self.zoneIdentifier) {
-    [GADMAdapterAppLovinUtils log: @"Invalid custom zone entered. Please double-check your credentials."];
-    NSError *error = [NSError errorWithDomain:GADMAdapterAppLovinConstant.errorDomain
-                                         code:0
-                                     userInfo:@{NSLocalizedFailureReasonErrorKey : @"Unable to resolve zone"}];
+  if (!_zoneIdentifier) {
+    [GADMAdapterAppLovinUtils
+        log:@"Invalid custom zone entered. Please double-check your credentials."];
+    NSError *error = GADMAdapterAppLovinErrorWithCodeAndDescription(kGADErrorMediationAdapterError,
+                                                                    @"Unable to resolve zone");
     [strongConnector adapter:self didFailAd:error];
-          
     return;
   }
-    
+
   [GADMAdapterAppLovinUtils log:@"Requesting interstitial for zone: %@", self.zoneIdentifier];
 
-  NSArray *requestedZones;
-  @synchronized(gRequestedInterstitialZoneIdentifiers) {
+  __block NSArray *requestedZones;
+  dispatch_sync(_lockQueue, ^{
     requestedZones = [NSArray arrayWithArray:gRequestedInterstitialZoneIdentifiers];
-  }
+  });
 
-  if ([requestedZones containsObject:self.zoneIdentifier]) {
-    [GADMAdapterAppLovinUtils log:@"Can't request a second ad for the same zone Identifier "
-                                  @"without showing the first ad."];
-    NSError *error = [NSError errorWithDomain:GADMAdapterAppLovinConstant.errorDomain
-                                         code:0
-                                     userInfo:nil];
+  if ([requestedZones containsObject:_zoneIdentifier]) {
```

Comment:
If a publishers has 1 AdMob ad unit set up, with only 1 AppLovin zone, if AdMob times out the AppLovin ad load, this will disqualify any future AppLovin ad load attempts... even if we successfully returned an ad in `adService:didLoadAd:`.

I understand that the goal of this logic is to prevent loading two AppLovin ads of the same zone for pubs who have back-to-back ads. Let me check internally to see if we can turn on a feature to allow back-to-back ads.

Reply:
If it times out we are removing that particular zone identifier at the dealloc time on line 108, which doesn't disqualify future requests.

Reply:
Ok there is a race condition there then, I believe [`dispatch_async(...`](https://github.com/googleads/googleads-mobile-ios-mediation/pull/149/files/5ec7d3f81b4564ba0b21b2f1eac8a4d1b2e60c7b#diff-78899a941343e15821a32b4276843661R106) should be `dispatch_sync`... Otherwise, `_interstitial = nil;` gets called first, causing the `if (self->_interstitial) {...}` check to always fail and zone to never be removed from the set.

Reply:
Added code to properly handle this case.